sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java

```diff
@@ -86,11 +88,39 @@ public void grow(int neededSize) {
     }
   }
 
-  public void reset() {
+  byte[] buffer() {
+    return buffer;
+  }
+
+  int getCursor() {
+    return cursor;
+  }
+
+  void incrementCursor(int val) {
+    cursor += val;
+  }
+
+  int pushCursor() {
```

Comment:
Can we make this a little bit less complex? I think just storing the cursor in the UnsafeWriter is enough.

Reply:
Since one `BufferHolder` is shared by multiple `UnsafeWriter`s, it seems to be simple to store cursors into `BufferHolders`.
WDYT?

Reply:
I also feel the current code is a bit complicated. Can't we avoid the sharing?

Reply:
Yeah, it is complicated. `pushCursor` will be dropped.